lib/logdna/client.rb

```diff
@@ -44,95 +50,125 @@ def process_message(msg, opts = {})
       processed_message
     end
 
-    def schedule_flush
-      start_timer = lambda {
-        sleep(@exception_flag ? @retry_timeout : @flush_interval)
-        flush if @flush_scheduled
-      }
-      Thread.new { start_timer.call }
+    def write_to_buffer(msg, opts)
+      processed_message = process_message(msg, opts)
+      message_size = processed_message.to_s.bytesize
+
+      buffer_byte_size = @lock.synchronize do
+        @buffer.push(Message.new(processed_message, message_size))
+        @buffer_byte_size += message_size
+      end
+
+      if buffer_byte_size >= @flush_threshold
+        @flush_scheduler.unschedule_flush
+        flush
+      else
+        @flush_scheduler.schedule_flush
+      end
     end
 
-    def write_to_buffer(msg, opts)
-      if @lock.try_lock
-        processed_message = process_message(msg, opts)
-        new_message_size = processed_message.to_s.bytesize
-        @buffer.push(processed_message)
-        @buffer_byte_size += new_message_size
-        @flush_scheduled = true
-        @lock.unlock
-
-        if @flush_limit <= @buffer_byte_size
-          flush
+    ##
+    # Asynchronously flush all logs in the buffer
+    def flush
+      Concurrent::Future.execute({ executor: @thread_pool }) { flush_sync }
+    end
+
+    ##
+    # Concurrently flush all logs in the buffer
+    def flush_sync
+      slices = @lock.synchronize do
+        # Minimize work in the critical section; figure out the slice to send and get out
+        if @buffer.any?
+          pop_points = []
+          pop_count = 0
+          running_size = 0
+
+          # Iterate backwards through array to avoid unnecessary memory allocation
+          @buffer.reverse_each.with_index do |message, index|
+            pop_count += 1
+            running_size += message.size
+            if running_size > @flush_limit || index == @buffer.length - 1
+              pop_points.push(pop_count)
+              @buffer_byte_size -= running_size
+              pop_count = 0
+              running_size = 0
+            end
+          end
+
+          slices = []
+          pop_points.each do |pop_point|
+            slices.push(@buffer.pop(pop_point).map { |message| message.source })
+          end
```

Comment:
I get what this whole section is doing, but imo it could use a block comment about how it's working. It's (at least to me) not obvious.  Or is there a simpler way to do the same thing?

Reply:
Yeah I think I can rewrite this with `Enumerable#reduce`. Might make it easier to read

Reply:
Simplified this logic and added a comment to explain what is being done and how.